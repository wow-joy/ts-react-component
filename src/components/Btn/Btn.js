import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import styled from "styled-components";

const Button = styled.span`
  height: 36px;
  padding: 0 20px;
  border: 1px solid transparent;
  font-size: 14px;
  cursor: pointer;
  display: inline-block;
  vertical-align: middle;
  line-height: 36px;
  user-select: none;
  ${props => props.defaultStyles};
`;
class BtnBase extends PureComponent {
  clickHandle = e => {
    const { onClick, to } = this.props;
    if (onClick && onClick(e) === false) {
      return;
    }
    if (to) {
      this.props.history.push(to);
    }
  };
  render() {
    const { defaultStyles, children, className } = this.props;
    return (
      <Button
        defaultStyles={defaultStyles}
        onClick={this.clickHandle}
        className={"wjc-btn " + className || ""}
      >
        {children}
      </Button>
    );
  }
}
class Btn extends PureComponent {
  render() {
    if (this.props.to) {
      const RouteBtn = withRouter(BtnBase);
      return <RouteBtn {...this.props} />;
    }
    return <BtnBase {...this.props} />;
  }
}
Btn.propTypes = {
  className: PropTypes.string,
  defaultStyles: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
  to: PropTypes.string
};
export default Btn;
