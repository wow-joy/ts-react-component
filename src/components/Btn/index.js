import Btn from './Btn';
import Hoc from "../../tools/Hoc";
import type1 from "./types/type1";
import type2 from "./types/type2";
import type3 from "./types/type3";
import disabled from "./types/disabled";

export const Type1 = Hoc(type1)(Btn);
export const Type2 = Hoc(type2)(Btn);
export const Type3 = Hoc(type3)(Btn);
export const Disabled = Hoc(disabled)(Btn);
export default Btn;