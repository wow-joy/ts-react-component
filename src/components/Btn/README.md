# Layout

按钮组件

## 使用

- 基础用法

```
import React from 'react'
import { Btn, Btn_1} from 'wowjoy-component'

const Foo = () => <Btn to={'/list'} > 跳转到list </Btn>
const Foo1 = () => <Btn_1 onClick={()=>console.log(1)} > 点击打印1 </Btn>
```

- Hoc 自定义默认状态

```
// file A
import React from 'react'
import { Btn }  from 'wowjoy-component/lib/components/Btn'
import { Hoc } from 'wowjoy-component/lib/tools'

const setting = {
  defaultStyles: `background: #06aea6;
    color: #fff;
    &:hover{
      background: #1AC3BB;
    }
    &:active{
      background: #3E8A86;
    }`
}

const A = () => Hoc(setting)(Btn)

//输出A为默认配置
export default A
```

```
// file B
import React from 'react'
import A from './A.js'

const B=()=><A onClick={()=>alert(1)}>点击alert1</A>
```

## APIs

| 属性          | 子属性 | 描述                                                                                                                                                               |   类型   | 默认值 |
| ------------- | ------ | :----------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------: | :----: |
| className     |        | 顶层样式 class                                                                                                                                                     |  string  |        |
| defaultStyles |        | 顶层默认样式                                                                                                                                                       |  string  |        |
| children      |        | 内容                                                                                                                                                               |   node   |        |
| to            |        | 路由跳转                                                                                                                                                           |  string  |        |
| onClick       |        | 点击事件。 <br>先于路由跳转执行，若要先执行路由跳转后执行点击事件，使用异步方式（_谨慎使用，异步方式不确定性太高_）<br>`return false` 会阻止`to`中设定的路由跳转。 | function |        |
