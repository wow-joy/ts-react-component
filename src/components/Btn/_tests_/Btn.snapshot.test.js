import React from "react";
import Btn from "../Btn.js";
import renderer from "react-test-renderer";
import * as TestUtils from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router-dom";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

const { renderIntoDocument } = TestUtils;

describe("Btn", () => {
  const reactObj = <Btn>Btn Content</Btn>;
  it("renders correctly", () => {
    const tree = renderer.create(reactObj).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should not use `to` outside a `Router`", () => {
    jest.spyOn(console, "error").mockImplementation(() => {});
    const reactObj = <Btn to="https://wow-joy.github.io">this is a link</Btn>;
    expect(()=>mount(reactObj)).toThrow(
      /You should not use <Route> or withRouter\(\) outside a <Router>/
    );
  });
  it("should render link when `to` has a value not equal false", () => {
    const reactDom = mount(
      <MemoryRouter>
        <Route pathname="/">
          <Btn to="https://wow-joy.github.io">this is a link</Btn>
        </Route>
      </MemoryRouter>
    );
    expect(reactDom.render()).toMatchSnapshot();
  });
  it("should not render as link when `to` equal false", () => {
    const reactObj = renderer.create(<Btn to={false}>this is not a link</Btn>);
    expect(reactObj.toJSON()).toMatchSnapshot();
  });
});
