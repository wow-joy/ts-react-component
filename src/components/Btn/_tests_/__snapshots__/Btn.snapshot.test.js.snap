// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Btn renders correctly 1`] = `
<span
  className="wjc-btn undefined sc-bdVaJa ezxaVl"
  onClick={[Function]}
>
  Btn Content
</span>
`;

exports[`Btn should not render as link when \`to\` equal false 1`] = `
<span
  className="wjc-btn undefined sc-bdVaJa ezxaVl"
  onClick={[Function]}
>
  this is not a link
</span>
`;

exports[`Btn should render link when \`to\` has a value not equal false 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "wjc-btn undefined sc-bdVaJa ezxaVl",
    },
    "children": Array [
      Object {
        "data": "this is a link",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "span",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
